tests/core/test_reader.py

```diff
@@ -201,6 +202,30 @@ def test_sensor_reader_read_one_raw(mock_sensor, sensor_reader_factory):
     assert raw_data.data
 
 
+def test_sensor_reader_read_one_warm_up(
+    mock_sensor,
+    sensor_reader_factory,
+    mock_sensor_warm_up,
+):
+    sensor_reader = sensor_reader_factory()
+
+    with sensor_reader as r:
+        with pytest.raises(pms.SensorNotReady):
+            sensor_reader.read_one()
+
+
+def test_sensor_reader_read_one_temp_failure(
+    mock_sensor,
+    sensor_reader_factory,
+    mock_sensor_temp_failure,
+):
+    sensor_reader = sensor_reader_factory()
+
+    with sensor_reader as r:
+        with pytest.raises(pms.SensorWarning):
+            sensor_reader.read_one()
+
+
```

Comment:
what is the purpose of `read_one`?
If it is to read one observation, these exceptions should be addressed internally and ask the sensor for an new message (silently).

If it is meant to request one observation (and let the caller handle the retry),
then this test are appropriate.

I think that `read_one` should deliver one observation, independently of how many samples are requested from the sensor and discarded for different reasons.

If you want the caller to handle the `retries`, I could add a `max_tries` option to `read_one` and `__call__`. This way when the `retry` exceeds `max_tries` the exception will bubble up to the caller.

Reply:
> If it is meant to request one observation (and let the caller handle the retry)

Yes, that's right: that's what I need it to do.

I had some questions, but then thought it would be better to share an example. I've appended some extra commits, the last two are what I think you're aiming for with `max_(re)tries` and `read_one`.

The last commit isn't great. I'd argue that's why `read_one` and ultimately `Stream`s are better, because it avoids confusion over the use of `samples` and `interval` (and `max_retries`) parameters. Given where we are, I think it would be clearer to drop the last commit and for me to just do `next(self())` manually.

Anyway, what do you think?

Reply:
Yes, this is what I had in mind. Thanks for the implementation.
I'll merge this PR as soon as I you confirm that you're done with it.

Thanks again,
Á.

Reply:
Should be ready for another look now...